core/common/src/main/scala/org/typelevel/otel4s/context/Context.scala

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2022 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s
+package context
+
+trait Context[C] { outer =>
+
+  /** The type of [[context.Key `Key`]] used by contexts of type `C`. */
+  type Key[A] <: context.Key[A]
+
+  /** Retrieves the value associated with the given key from the context, if
+    * such a value exists.
+    */
+  def get[A](ctx: C)(key: Key[A]): Option[A]
+
+  /** Retrieves the value associated with the given key from the context, if
+    * such a value exists; otherwise, returns the provided default value.
+    */
+  def getOrElse[A](ctx: C)(key: Key[A], default: => A): A =
+    get(ctx)(key).getOrElse(default)
+
+  /** Creates a copy of this context with the given value associated with the
+    * given key.
+    */
+  def updated[A](ctx: C)(key: Key[A], value: A): C
+
+  /** The root context, from which all other contexts are derived. */
+  def root: C
+
+  class Ops(ctx: C) {
```

Comment:
Should the class itself be final and the constructor be private? 
```scala
final class Ops private[Context] (ctx: C)
```

Reply:
it is modeled after [`Ordering#OrderingOps`](https://github.com/scala/scala/blob/v2.13.11/src/library/scala/math/Ordering.scala#L192-L200) from the stdlib, for which neither of those are the case. it allows one to theoretically extend `Context` with more methods and have a new inner class extend `Ops` as well to add additional methods. is that use case significant enough? I don't know.

it does remind me however why it's called `OrderingOps` and not just `Ops` (so that there isn't a naming conflict with a child `Ops` class), so I'll tweak that at least